src/main/java/com/microsoft/azure/util/AzureCredentials.java

```diff
@@ -33,30 +38,78 @@ public ValidationException(String message) {
-                = "https://management.azure.com/";
-        public static final String DEFAULT_GRAPH_ENDPOINT
-                = "https://graph.windows.net/";
-        public static final String DEFAULT_OAUTH_PREFIX
-                = "https://login.windows.net/<TenantId>";
+        static final String ENV_AZURE = "Azure";
+        static final String ENV_AZURE_CHINA = "Azure China";
+        static final String ENV_AZURE_GERMANY = "Azure Germany";
+        static final String ENV_AZURE_US_GOVERNMENT = "Azure US Government";
     }
 
     public static class ServicePrincipal implements java.io.Serializable {
+        private static final long serialVersionUID = 1L;
 
         private final Secret subscriptionId;
         private final Secret clientId;
         private final Secret clientSecret;
-        private final Secret oauth2TokenEndpoint; //keeping this for backwards compatibility
-        private final String serviceManagementURL;
-        private final Secret tenant;
-        private final String authenticationEndpoint;
-        private final String resourceManagerEndpoint;
-        private final String graphEndpoint;
+        private Secret oauth2TokenEndpoint; //keeping this for backwards compatibility
+        private String serviceManagementURL;
+        private Secret tenant;
+        private String authenticationEndpoint;
+        private String resourceManagerEndpoint;
+        private String graphEndpoint;
+
+        /**
+         * Name of the Azure Environment.
+         * <p>
+         * Added in the migration to Azure SDK 1.3.0.
+         */
+        private String azureEnvironmentName;
+
+        /**
+         * Cache of the resolved azure environment.
+         * <p>
+         * This should be cleared whenever the {@link #azureEnvironmentName}, or any of the endpoint override
+         * is updated.
+         */
+        private transient AzureEnvironment azureEnvironment;
+
+        /**
+         * After deserialization hook to upgrade legacy service principal data.
+         * <p>
+         * XStream serialization / deserialization used by Jenkins doesn't support objects with  readObject/writeObject
+         * defined.
+         */
+        private Object readResolve() throws ObjectStreamException {
+            Map<String, AzureEnvironment> environmentMap = new HashMap<>();
+            environmentMap.put(Constants.ENV_AZURE, AzureEnvironment.AZURE);
+            environmentMap.put(Constants.ENV_AZURE_CHINA, AzureEnvironment.AZURE_CHINA);
+            environmentMap.put(Constants.ENV_AZURE_GERMANY, AzureEnvironment.AZURE_GERMANY);
+            environmentMap.put(Constants.ENV_AZURE_US_GOVERNMENT, AzureEnvironment.AZURE_US_GOVERNMENT);
+
+            // If the environment name is not recognized, which may happen when the user upgraded the plugin
+            // and didn't update the credentials, we try to match a known environment.
+            boolean matched = false;
```

Comment:
Should check `azureEnvironmentName` first?

Reply:
Yes. It was missed when I did the code refactoring. Thanks for the catch, I will fix this.

Reply:
added the condition check